src/corehost/cli/json_parser.cpp

```diff
@@ -0,0 +1,130 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// These are only used by rapidjson/error/en.h to declare the error messages,
+// and have to be set to these values before any files are included.  They're
+// defined here because it's the only place that calls GetParseError().
+#undef RAPIDJSON_ERROR_CHARTYPE
+#undef RAPIDJSON_ERROR_STRING
+#define RAPIDJSON_ERROR_CHARTYPE pal::char_t
+#define RAPIDJSON_ERROR_STRING(x) _X(x)
+
+#include "json_parser.h"
+#include "rapidjson/error/en.h"
+#include "utils.h"
+#include <cassert>
+#include <cstdint>
+
+namespace {
+
+// Try to match 0xEF 0xBB 0xBF byte sequence (no endianness here.)
+std::streampos get_utf8_bom_length(pal::istream_t& stream)
+{
+    if (stream.eof())
+    {
+        return 0;
+    }
+
+    auto peeked = stream.peek();
+    if (peeked == EOF || ((peeked & 0xFF) != 0xEF))
+    {
+        return 0;
+    }
+
+    unsigned char bytes[3];
+    stream.read(reinterpret_cast<char*>(bytes), 3);
+    if ((stream.gcount() < 3) || (bytes[1] != 0xBB) || (bytes[2] != 0xBF))
+    {
+        return 0;
+    }
+
+    return 3;
+}
+
+void get_line_column_from_offset(const std::vector<char>& json, size_t offset, int *line, int *column)
+{
+    *line = *column = 1;
+
+    for (size_t i = 0; i < offset; i++)
+    {
+        (*column)++;
+
+        if (json[i] == '\n')
+        {
+            (*line)++;
+            *column = 1;
+        }
+        else if (json[i] == '\r' && json[i + 1] == '\n')
+        {
+            (*line)++;
+            *column = 1;
+
+            i++; // Discard carriage return
+        }
+    }
+}
+
+} // empty namespace
+
+void json_parser_t::realloc_buffer(size_t size)
+{
+    m_json.resize(size + 1);
+    m_json[size] = '\0';
+}
+
+bool json_parser_t::parse_json(const pal::string_t& context)
+{
+    assert(!m_json.empty());
+
+#ifdef _WIN32
+    // Can't use in-situ parsing on Windows, as JSON data is encoded in
+    // UTF-8 and the host expects wide strings.  m_document will store
+    // data in UTF-16 (with pal::char_t as the character type), but it
+    // has to know that data is encoded in UTF-8 to convert during parsing.
+    m_document.Parse<rapidjson::ParseFlag::kParseStopWhenDoneFlag, rapidjson::UTF8<>>(m_json.data());
+#else
+    m_document.ParseInsitu(m_json.data());
+#endif
+
+    if (m_document.HasParseError())
+    {
+        int line, column;
+        size_t offset = m_document.GetErrorOffset();
+
+        get_line_column_from_offset(m_json, offset, &line, &column);
```

Comment:
I was going to ask if we wanted to file something in rapidjson suggesting supporting this and found they already had [916](https://github.com/Tencent/rapidjson/issues/916). Is the CursorStreamWrapper that was added something we could use to figure out the line/column?

Reply:
Yeah, I found the same issue in their repo; that's why I wrote the function.  I haven't looked at CursorStreamWrapper yet; will see if it works too.

Reply:
From a first glance, it doesn't seem I can use in-situ parsing when using CursorStreamWrapper; it wraps an InputStream whereas the in-situ parsing mode expects a character array compatible with the input encoding type.

Even if that's not a problem, since errors should be rare occurrences, I'd rather not add the (tiny) overhead of counting lines and columns except if absolutely necessary.

Also, it doesn't handle `\r\n` line endings, only `\n`.  (Not a big deal, as columns would be off by 1, which would be good enough in most cases to determine the location of an error, but still...)